src/prefect/tasks/shell.py

```diff
@@ -110,6 +110,7 @@ def run(self, command: str = None, env: dict = None) -> str:
                     # if we're returning all, we don't log every line
                     self.logger.debug(line)
             sub_process.wait()
+            sub_process.stdout.close()
```

Comment:
A different way to write this would be to use `Popen` as a contextmanager - in that case all resources are ensured cleaned up before exiting. Currently the `stdout.close()` here will only close if no error occurs above, which can't be guaranteed. Something like:

```python
with Popen(
    [self.shell, tmp.name], stdout=PIPE, stderr=STDOUT, env=current_env
) as sub_process:
    lines = []
    line = None
    for raw_line in iter(sub_process.stdout.readline, b""):
        line = raw_line.decode("utf-8").rstrip()
        if self.return_all:
            lines.append(line)
        else:
            # if we're returning all, we don't log every line
            self.logger.debug(line)
    sub_process.wait()
```

Reply:
fixed in 224e179 :+1: